File: src/main/scala/com/hoi4utils/clausewitz/script/BooleanPDX.scala
package com.hoi4utils.clausewitz.script

import com.hoi4utils.clausewitz.BoolType
import com.hoi4utils.clausewitz_parser.Node
import com.hoi4utils.clausewitz_parser.NodeValue
import org.jetbrains.annotations.NotNull

class BooleanPDX(pdxIdentifiers: List[String], final private var defaultValue: Boolean, final private var boolType: BoolType) extends AbstractPDX[Boolean](pdxIdentifiers) {

  def this(pdxIdentifier: String, defaultValue: Boolean, boolType: BoolType) = {
    this(List(pdxIdentifier), defaultValue, boolType)
  }

  @throws[UnexpectedIdentifierException]
  @throws[NodeValueTypeException]
  override def set(expression: Node): Unit = {
    usingIdentifier(expression)
    this.node = Some(expression)
    expression.$ match {
      case _: Boolean =>
//      case _: String =>
//        val v = this.node.$.asInstanceOf[String]
//        if (boolType.maches(v)) obj = boolType.parse(v)
//        else throw new NodeValueTypeException(expression, "String parsable as Bool matching enum + " + boolType.toString)
      case _ => throw new NodeValueTypeException(expression, "Boolean or String")
    }
  }
  
  override def set(value: Boolean): Unit = {
    if (this.node.nonEmpty)
      this.node.foreach(_.setValue(value))
    else
      this.node = Some(Node(NodeValue(value)))
  }

  override def get(): Option[Boolean] = {
    val v = super.get()
    v.get match {
      case b: Boolean => Some(b)
      case _ => None
    }
  }

  def $ : Boolean = {
    get().getOrElse(defaultValue)
  }

  def objEquals(other: PDXScript[?]): Boolean = {
    other match {
      case b: BooleanPDX => node.equals(b.node)
      case _ => false
    }
  }

  def invert: Boolean = {
    setNode(!this.$)
    this.$
  }
}

File: src/main/java/com/hoi4utils/clausewitz/script/BooleanPDX.java
package com.hoi4utils.clausewitz.script;

import com.hoi4utils.clausewitz.BoolType;
import com.hoi4utils.clausewitz_parser.Node;
import com.hoi4utils.clausewitz_parser.NodeValue;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class BooleanPDX extends AbstractPDX<Boolean> {
    private final BoolType boolType;
    /**
     * this is separate to obj because
     * if we want to clear the value of obj, we would still return this value by default.
     */
    private final boolean defaultValue;

    public BooleanPDX(String pdxIdentifiers, boolean defaultValue, BoolType boolType) {
        super(pdxIdentifiers);
        this.boolType = boolType;
        this.defaultValue = defaultValue;
    }

    public BooleanPDX(List<String> pdxIdentifiers, boolean defaultValue, BoolType boolType) {
        super(pdxIdentifiers);
        this.boolType = boolType;
        this.defaultValue = defaultValue;
    }

    @Override
    public void set(Node expression) throws UnexpectedIdentifierException, NodeValueTypeException {
        usingIdentifier(expression);
        NodeValue value = expression.value();

        if (value.valueObject() instanceof String str) {
            obj = value.bool(boolType);
        } else {
            throw new NodeValueTypeException(expression, "String parsable as Bool matching enum + " + boolType.toString());
        }
    }

    @Override
    public @NotNull Boolean get() {
        var val = super.get();
        if (val == null) {
            return defaultValue;
        }
        return val;
    }

    @Override
    public boolean objEquals(PDXScript<?> other) {
        if (other instanceof BooleanPDX pdx) {
            return obj.equals(pdx.get());
        }
        return false;
    }

    public boolean invert() {
        set(!get());
        return get();
    }
}
